File: backends/Contracts/MemoryCache.Default/Handlers/Remove.cs
using D2.Contracts.Handler;
using D2.Contracts.Result;
using Microsoft.Extensions.Caching.Memory;
using H = D2.Contracts.Interfaces.ICommonCacheService.IRemoveHandler;
using I = D2.Contracts.Interfaces.ICommonCacheService.RemoveInput;
using O = D2.Contracts.Interfaces.ICommonCacheService.RemoveOutput;

namespace D2.Contracts.MemoryCache.Default.Handlers;

/// <inheritdoc cref="H"/>
public class Remove : BaseHandler<H, I, O>, H
{
    /// <summary>
    /// Initializes a new instance of the <see cref="Remove"/> class.
    /// </summary>
    ///
    /// <inheritdoc/>
    public Remove(
        IMemoryCache memoryCache,
        IHandlerContext context) : base(context)
    {
        r_memoryCache = memoryCache;
    }

    private readonly IMemoryCache r_memoryCache;

    /// <inheritdoc/>
    protected override ValueTask<D2Result<O?>> ExecuteAsync(
        I input,
        CancellationToken ct = default)
    {
        r_memoryCache.Remove(input.Key);
        return ValueTask.FromResult(D2Result<O?>.Ok(new O(), traceId: TraceId));
    }
}

File: backends/node/shared/implementations/caching/in-memory/default/src/handlers/d/remove.ts
import { BaseHandler, type IHandlerContext } from "@d2/handler";
import { D2Result } from "@d2/result";
import type { InMemoryCache } from "@d2/interfaces";
import type { MemoryCacheStore } from "../../memory-cache-store.js";

type Input = InMemoryCache.RemoveInput;
type Output = InMemoryCache.RemoveOutput;

export class Remove extends BaseHandler<Input, Output> implements InMemoryCache.IRemoveHandler {
  private readonly store: MemoryCacheStore;

  constructor(store: MemoryCacheStore, context: IHandlerContext) {
    super(context);
    this.store = store;
  }

  protected async executeAsync(input: Input): Promise<D2Result<Output | undefined>> {
    this.store.delete(input.key);
    return D2Result.ok({ data: {}, traceId: this.traceId });
  }
}
